### PR TITLE
add Pharo13 support

### PR DESCRIPTION
**Pull request by estebanlm**
we are about to release Pharo12, which means we need to open Pharo13 :)

### CODE REVIEW

**Comment by theseion** [comment]
Looks good, except for the 404 from the file server.

**Comment by estebanlm** [comment]
yes, I made the PR before having the images :P
can you re-trigger it ?

**Comment by theseion** [comment]
```
Preparing Pharo image...
Unsupported Pharo version 'Pharo64-13'.
Downloading Pharo64-13 vm...
download_file() expects an URL and a target path.
Error: Process completed with exit code 1.
```

**Comment by fniephaus** [comment]
> Unsupported Pharo version 'Pharo64-13'.

@estebanlm please also update `pharo::get_vm_url()`

**Comment by estebanlm** [comment]
I have no idea what is happening now :)

**Comment by theseion** [comment]
The specification is good now, IMO. I need to look into the failures coming from SmalltalkCI.

**Comment by theseion** [comment]
Some of the tests fail due to issues with package management, even in Pharo12-stable. That's odd because I had fixed all of that once already. @estebanlm I'd appreciate it if you could help out with those test failures (or @jecisc maybe). From a quick glance it looks like there's some confusion of tags / categories.

**Comment by theseion** [comment]
Nvm, I figured it out. @fniephaus the last failing test (`#testReportFailure`) fails due to different ordering of tests, which leads to different output strings. Also, the fixtures appear to be outdated. I don't see a good way to make that test stable to be honest, the fact that it has worked for so long is a miracle...

**Comment by theseion** [comment]
I spent 30 minutes trying to clean up the stdout reporter tests and then gave up. To properly match the generated output, we'd either have to use a custom runner or implement templating and complex matching. I don't think it's worth the effort. I've removed the test.

**Comment by theseion** [comment]
Currently, the GemStone builds are failing and I can't debug those. @fniephaus, can  you help out?

**Comment by fniephaus** [comment]
> I spent 30 minutes trying to clean up the stdout reporter tests and then gave up. To properly match the generated output, we'd either have to use a custom runner or implement templating and complex matching. I don't think it's worth the effort. I've removed the test.

Could we relax the test, for example, by ensuring every line from the template can be found in the generated output?

**Comment by fniephaus** [comment]
> Currently, the GemStone builds are failing and I can't debug those. @fniephaus, can you help out?

It seems [`hdiutil` is failing for some reason](https://github.com/hpi-swa/smalltalkCI/actions/runs/8889479714/job/24407948184?pr=639#step:4:428):
```
UserDefinedError: The command '/usr/bin/hdiutil attach GemStone64Bit3.5.3-arm64.Darwin.dmg' exited with status 1. See the file '/tmp/gsd_stones139996781290505153618979304401482798518.err'
```

Maybe the `.err` file would provide us with more details but this really is something @dalehenrich needs to help us with. In the meantime, I think it's ok to mark the failing GemStone builds as allowed to fail.

**Comment by theseion** [comment]
> Could we relax the test, for example, by ensuring every line from the template can be found in the generated output?

I tried that. Unfortunately, the output includes stuff that I'm sure will differ on other platforms. Example:
```

[1m[34m#################################################
[1m[34m# Stdout-testReportFailure                      #
[1m[34m# 8 Tests with 4 Failures and 1 Errors in 0.02s #
[1m[34m#################################################
[0m
(3 tests passed)
[1m
SCIExcludedTests
[0m[32m ✓[0m #testDeprecation (0ms)[0m
[32m ✓[0m #testShouldPass (0ms)[0m
[32m ✓[0m #testShouldFail (1ms)[0m

[1m[31m#########################
[1m[31m# 5 tests did not pass: #
[1m[31m#########################
[0m[1m
SCIExcludedTests
[0m[1m[33m ✗ #testThisIsAVeryLongMethodNameThat...playedCorrectlyInATravisLog (7ms)[0m
[1mTestFailure: Assertion failed[0m
SCIExcludedTests(TestAsserter)>>assert:description:resumable:
SCIExcludedTests(TestAsserter)>>assert:description:
SCIExcludedTests(Object)>>assert:
SCIExcludedTests>>testThisIsAVeryLongMethodNameThatProbablyNeedsToBeContractedInOrderToBeDisplayedCorrectlyInATravisLog ...assert: false
SCIExcludedTests(TestCase)>>performTest
[0m[1m[33m ✗ #testShouldPassUnexpectedly (0ms)[0m
[1mTestFailure: Test passed unexpectedly[0m[1m[31m ✗ #testError (1ms)[0m
[1mError: An error message.[0m
SCIExcludedTests(Object)>>error:
SCIExcludedTests>>testError ...error: 'An error message.'
SCIExcludedTests(TestCase)>>performTest
[0m[1m[33m ✗ #testFailure (1ms)[0m
[1mTestFailure: A failure message.[0m
SCIExcludedTests(TestAsserter)>>assert:description:resumable:
SCIExcludedTests(TestAsserter)>>assert:description:
SCIExcludedTests(TestAsserter)>>fail:
SCIExcludedTests>>testFailure ...fail: 'A failure message.'
SCIExcludedTests(TestCase)>>performTest
[0m[1m[33m ✗ #testAssertError (1ms)[0m
[1mTestFailure: Got 3 instead of 4.[0m
SCIExcludedTests(TestAsserter)>>assert:description:resumable:
SCIExcludedTests(TestAsserter)>>assert:description:
SCIExcludedTests(TestAsserter)>>assert:equals:
SCIExcludedTests>>testAssertError ...assert: 3 equals: 4
SCIExcludedTests(TestCase)>>performTest
[0m

[1m[31m###########
[1m[31m# Summary #
[1m[31m###########
[0m[1m
SCIExcludedTests
[0m[33m ✗ #testThisIsAVeryLongMethodNameThatProbablyNeedsToBeContractedInOrderToBeDisplayedCorrectlyInATravisLog (7ms)[0m
[33m ✗ #testShouldPassUnexpectedly (0ms)[0m
[31m ✗ #testError (1ms)[0m
[33m ✗ #testFailure (1ms)[0m
[33m ✗ #testAssertError (1ms)[0m

[1m[31msmalltalkCI Deprecation Warnings[0m
 - SCIExcludedTests>>testDeprecation (This is just a test)


[1m[31m  Executed 8 Tests with 4 Failures and 1 Errors in 0.02s.[0m


```

Fixture:
```
failureFixtureNonTravis
	^ '
[1m[34m###############
[1m[34m# Stdout-testReportFailure#
[1m[34m# 8 Tests with 4 Failures and 1 Errors in s #
[1m[34m###############
[0m
(3 tests passed)
[1m
SCIExcludedTests
[0m[32m ✓[0m #testDeprecation (ms)[0m
[32m ✓[0m #testShouldFail (ms)[0m
[32m ✓[0m #testShouldPass (ms)[0m

[1m[31m#########################
[1m[31m# 5 tests did not pass: #
[1m[31m#########################
[0m[1m
SCIExcludedTests
[0m[1m[33m ✗ #testAssertError (ms)[0m
[0m[1m[31m ✗ #testError (ms)[0m
[0m[1m[33m ✗ #testFailure (ms)[0m
[0m[1m[33m ✗ #testShouldPassUnexpectedly (ms)[0m
[0m[1m[33m ✗ #testThisIsAVeryLongMethodNameThat...playedCorrectlyInATravisLog (ms)[0m

[1m[31m###########
[1m[31m# Summary #
[1m[31m###########
[0m[1m
SCIExcludedTests
[0m[33m ✗ #testAssertError (ms)[0m
[31m ✗ #testError (ms)[0m
[33m ✗ #testFailure (ms)[0m
[33m ✗ #testShouldPassUnexpectedly (ms)[0m
[33m ✗ #testThisIsAVeryLongMethodNameThatProbablyNeedsToBeContractedInOrderToBeDisplayedCorrectlyInATravisLog (ms)[0m
[1m[31msmalltalkCI Deprecation Warnings[0m
 - SCIExcludedTests>>testDeprecation (This is just a test)


[1m[31m  Executed 8 Tests with 4 Failures and 1 Errors in s.[0m

'
```

As you can see, the current result includes stack information and other strings that are not part of the fixture.

**Comment by estebanlm** [comment]
I do not understand. Anything I should do?

**Comment by theseion** [comment]
No, I need a decision from @fniephaus on how to proceed.

**Comment by fniephaus** [comment]
> As you can see, the current result includes stack information and other strings that are not part of the fixture.

Sure, makes sense. Could the fixture just be an order-independent subset of the expected output? Something like a list of substrings that must be found in the output.

**Comment by theseion** [comment]
That sounds feasible, yes.

**Comment by fniephaus** [comment]
Hope it's not too much work, otherwise I'm ok with disabling/removing the test.

**Comment by theseion** [comment]
I have a working solution. Can you get me the current output for Travis? Or should we just skip the test for Travis?

**Comment by theseion** [comment]
@fniephaus I give up. I had it and then GemStone got in the way with a failing implementation of `String>>#match:` (possibly because of `WideString`) before I discovered that in Pharo 12 sometimes a new line is missing. There's just too much that can go wrong. IMO, at this point the test would be pretty much useless because it has to incorporate too many exceptions. I've removed it now.

**Comment by fniephaus** [comment]
Ok, thanks for trying anyway! So this is good to go now from your perspective?

File: repository/SmalltalkCI-Pharo-Coverage-Core.package/SCIPharo13CodeCoverage.class/properties.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+	"commentStamp" : "",
+	"super" : "SCIPharoCodeCoverage",
+	"category" : "SmalltalkCI-Pharo-Coverage-Core",
+	"classinstvars" : [ ],
+	"pools" : [ ],
+	"classvars" : [ ],
+	"instvars" : [ ],
+	"name" : "SCIPharo13CodeCoverage",
```
Comment (fniephaus) [inline repository/SmalltalkCI-Pharo-Coverage-Core.package/SCIPharo13CodeCoverage.class/properties.json:9]: This class does not have any methods, right? Why is it needed or can it be removed?

Comment (estebanlm) [reply repository/SmalltalkCI-Pharo-Coverage-Core.package/SCIPharo13CodeCoverage.class/properties.json:9]: I did not create that class :)

Comment (theseion) [reply repository/SmalltalkCI-Pharo-Coverage-Core.package/SCIPharo13CodeCoverage.class/properties.json]: That was me. It's not required. I've removed it.

**Comment by theseion** [comment]
> Ok, thanks for trying anyway! So this is good to go now from your perspective?

Yes, looks good (with the exception of the GemStone macOS builds).

**Comment by theseion** [comment]
I'll squash before merging.

**Comment by theseion** [comment]
The Pharo64-6.1 on windows-2019 job failed to upload coverage results (the job didn't fail). Do you consider that an issue @fniephaus?

```
Uploading coverage results to Coveralls...
Failed to upload coverage results (HTTP status code #500):
{"message":"Build processing error.","error":true,"url":""}
```

**Comment by fniephaus** [comment]
If it fails the build, I guess it's a bug. But nothing that needs to be addressed in this PR. Don't want to delay this any further :)

**Review by fniephaus — APPROVED** [review]
LGTM

**Comment by theseion** [comment]
Now it looks good. I had to adjust the timeout for Squeak trunk on macOS. For some reason, each stop / start cycle of the images takes multiple minutes.

I can't merge btw, don't have the permission.

**Comment by fniephaus** [comment]
> I can't merge btw, don't have the permission.

That's weird, you do have write permissions which should include permission to merge PRs.

**Comment by fniephaus** [comment]
Oh, did we just deploy this on a Friday :crossed_fingers: :wink:

**Comment by theseion** [comment]
IT'S GOING TO BE FINE! 😅

**Comment by jecisc** [comment]
I wonder if something is missing because we still have this in the CI:

```
Starting Pharo build...
Unsupported Pharo version 'Pharo64-13'.
Downloading Pharo64-13 image...
download_file() expects an URL and a target path.
Error: Process completed with exit code 1.
```

**Comment by theseion** [comment]
Where are you seeing this?

**Comment by jecisc** [comment]
In most external projects of Pharo. Example:

https://github.com/pharo-spec/NewTools/pull/753/checks

**Comment by theseion** [comment]
That's because there is no `Pharo64-13`. Use `Pharo64-alpha` instead. This is a deliberate choice. With Pharo 12, the image at `.../120` was always stale. So while the expectation was that `Pharo64-alpha` = `Pharo64-12`, this wasn't true and led to some weird issues. We will add `Pharo64-13` before it is released.

**Comment by estebanlm** [comment]
> That's because there is no `Pharo64-13`. Use `Pharo64-alpha` instead. This is a deliberate choice. With Pharo 12, the image at `.../120` was always stale. So while the expectation was that `Pharo64-alpha` = `Pharo64-12`, this wasn't true and led to some weird issues. We will add `Pharo64-13` before it is released.

I am sorry but I disagree with that choice. If ../120 (get-files?) was stale there was a problem there, but I do not see why we need to remove the naming for that (which poses some problems in our flow, also).

**Comment by theseion** [comment]
Ok, I'll add entries for Pharo 13. I'll be counting on your help when stuff breaks ;)

**Pull request merged**